spotbugs/src/main/java/edu/umd/cs/findbugs/detect/DontAssertInstanceofInTests.java

```diff
@@ -0,0 +1,76 @@
+package edu.umd.cs.findbugs.detect;
+
+import edu.umd.cs.findbugs.BugInstance;
+import edu.umd.cs.findbugs.BugReporter;
+import edu.umd.cs.findbugs.BytecodeScanningDetector;
+
+import org.apache.bcel.Const;
+import org.apache.bcel.classfile.AnnotationEntry;
+import org.apache.bcel.classfile.Code;
+import org.apache.bcel.classfile.Method;
+
+/**
+ * A detector that checks for lines in JUnit tests that look like `assertTrue(object instanceof Class)` and discourages them.
+ *
+ * It may be more useful to observe error messages from bad casts, as this may reveal more information regarding the
+ * circumstances of the exception than a "false is not true" assert message.
+ *
+ * This detector reports bugs of the type `JUA_DONT_ASSERT_INSTANCEOF_IN_TESTS`.
+ */
+public class DontAssertInstanceofInTests extends BytecodeScanningDetector {
+    private boolean isTest;
+
+    private BugInstance currBug = null;
+
+    BugReporter bugReporter;
+
+    public DontAssertInstanceofInTests(BugReporter bugReporter) {
+        this.bugReporter = bugReporter;
+    }
+
+    @Override
+    public void visitMethod(Method obj) {
+        isTest = false;
+        for (AnnotationEntry entry : obj.getAnnotationEntries()) {
+            String entryType = entry.getAnnotationType();
+            isTest |= "Lorg/junit/Test;".equals(entryType);
+        }
+    }
+
+    @Override
+    public void visitCode(Code obj) {
+        if (isTest)
+            super.visitCode(obj);
```

Comment:
If the purpose is skipping `sawOpcode(int)`, it could be better to implement [`beforeOpcode(int)`](https://javadoc.io/static/com.github.spotbugs/spotbugs/4.1.1/edu/umd/cs/findbugs/visitclass/DismantleBytecode.html#beforeOpcode-int-) instead.

Reply:
does `beforeOpcode` control if we want to skip processing an entire method? The intention was to avoid processing any methods that weren't annotated as tests.

Reply:
I think so. The javadoc describes its responsibility as `return false if we should skip calling sawOpcode`, and the implementation is like below:

https://github.com/spotbugs/spotbugs/blob/c195f7e6632c078bcb8bbf8d3036476d5956f2f1/spotbugs/src/main/java/edu/umd/cs/findbugs/visitclass/DismantleBytecode.java#L877-L879

Reply:
from my understanding, beforeOpcode would run on every opcode of a method.
Now that I look, there's a `shouldVisitCode` method in BytecodeScanningDetector which would do exactly what I was trying to do, so I'll just override that instead of that `visitCode` hack.